File: src/test/java/com/uapp_llc/test/model/factory/ModelFactory.java
package com.uapp_llc.test.model.factory;

import com.uapp_llc.test.model.type.ModelType;
import com.uapp_llc.test.model.wrapper.ModelWrapper;

public final class ModelFactory {

  private ModelFactory() {
  }

  public static <T> ModelWrapper<T> createWrapper(ModelType<T> type) {
    Class<T> modelClass = type.modelClass();
    AbstractFactory<T> factory = FactoryProducer.getFactory(modelClass);
    return factory.createWrapper(type);
  }

  public static <T> T createModel(ModelType<T> type) {
    ModelWrapper<T> wrapper = ModelFactory.createWrapper(type);
    return (wrapper == null) ? null : wrapper.getModel();
  }

}

File: src/test/kotlin/com/uapp_llc/test/model/factory/ModelFactory.kt
package com.uapp_llc.test.model.factory

import com.uapp_llc.test.model.type.ModelType

object ModelFactory {

  @JvmStatic
  fun <T> createModel(type: ModelType<T>): T {
    val modelClass: Class<T> = type.modelClass()
    return FactoryProducer.getFactory(modelClass).createModel(type)
  }

}
